core/src/plugins/container/cloudbuilder.ts

```diff
@@ -55,20 +56,140 @@ const cloudBuilderAvailability = new LRUCache<string, CloudBuilderAvailabilityV2
   ttl: 1000 * 60 * 5,
 })
 
+type RetrieveAvailabilityParams = {
+  ctx: PluginContext
+  action: Resolved<ContainerBuildAction>
+  config: CloudBuilderConfiguration
+}
+
+async function retrieveAvailabilityFromCloud(params: {
+  ctx: PluginContext
+  action: Resolved<ContainerBuildAction>
+  config: CloudBuilderConfiguration
+}): Promise<CloudBuilderAvailabilityV2> {
+  if (gardenEnv.USE_GARDEN_CLOUD_V2) {
+    return await retrieveAvailabilityFromGrowCloud(params)
+  } else {
+    return await retrieveAvailabilityFromGardenCloud(params)
+  }
+}
+
+async function retrieveAvailabilityFromGardenCloud({
+  ctx,
+  action,
+  config,
+}: RetrieveAvailabilityParams): Promise<CloudBuilderAvailabilityV2> {
+  const { isInClusterBuildingConfigured } = config
+  if (!ctx.cloudApi) {
```

Comment:
Can we somehow make sure we don't duplicate this if statement, perhaps by moving it to the caller of `retrieveAvailability` and use this if statement:
```suggestion
  if (!ctx.cloudApi && !ctx.cloudApiV2) {
```

Reply:
I extracted a helper function to prepare an error and to reduce code duplication in bdc03692eaa23b26061e6f06c5fabb04c1b18d1e.

Semantically, the login status check was the part of each backend-specific function, and it still can/should be.
If we pull these checks up, then in each backend-specific function we'll have to perform undefined-checks for `ctx.cloudApi` and `ctx.cloudApiV2` again, or assume those are respectively defined by using `!` operator.

The latter way is brittle, the former could be simplified if we introduce predicate function and guard-types `LoggedInPluginContext` and `LoggedInPluginContextV2` for `PluginContext` (like we introduced `LoggedInGarden` for `Garden`).

That would be more type-safe, but requires more changes.
We can improve it later in the follow-up PRs to unlock this and git it merged to the main asap.

Wdyt?

Reply:
Yes, we can have a kind of template function there, and make the backend-specific behavior abstract and parameterizable. Working on it now.

Reply:
Done, pls check the last few commits.